### PR TITLE
adds permissions data job to workflow

### LINKED ISSUE #690: Gather permissions data for publications
We'll need an automated step after file version is obtained to gather permissions data from the OA Works permissions API.  We should try to share as much code as possible with the OA Works permissions integration that already exists.  We'll need a `licence`, `embargo_date`, and `set_statement` field added to the publications table for this.  I don't think there needs to be a manual stop immediately for this.  If there is permissions data missing, then I think this can be addressed at the metadata curation stage?

We'll also need to use the permissions API and the versions of the files to determine if we have a correct version.  We should add a field to `Publication` that stores the preferred version ('accepted' or 'published').  If this check finds a file that is the correct version, then it will be good to move to the next step.  I don't think we need to delete the other files.  We may just want to hide them from the admins or just make sure they are labeled as the wrong version and then wait to delete any unused files at the end of the whole workflow. 

If the proper version is not found in any of the attached files, then they will all be marked as not the proper version.  They will then show up in the GUI for manual curation.  I don't think we need to set a workflow status for this.  It should work like the DOI check.

### PR DESCRIPTION
**Pull request by Smullz622**
Fixes #690

### CODE REVIEW

**Review by EricDurante — COMMENTED** [review]
Depending on what we decide with regard to the comments that I already posted, we may need to change some things, so I'm going to hold off on reviewing the rest of this until we have some discussion.

File: app/jobs/permissions_check_job.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+class PermissionsCheckJob < ApplicationJob
+  queue_as 'default'
+
+  def perform(file_id)
+    file = ActivityInsightOaFile.find(file_id)
+    publication = Publication.find(file.publication_id)
+    permissions = OabBestPermissionsService.new(publication.doi_url_path)
+    publication.preferred_version = permissions.best_version
+
+    if permissions.best_version == file.version
```
[the remark this answers is not in the document]
Comment (ajkiessl) [reply app/jobs/permissions_check_job.rb:12]: That sounds like it's on the right track and similar to what we do with the permissions check during the Scholarsphere upload.

So, maybe what we can do is loop through all the permissions we get back from the API.  If there is an Accepted Version that allows archival and has "Institutional Repository " as it's location _and_ there is no Published Version, then we store Accepted Version as it's preferred version and take it's metadata.  If there is no Accepted Version and only a Published Version in the returned permissions, then we store Published Version as the preferred version and take it's metadata.

I'm not too sure what would happen if both of the former criteria are true though.  Looking through a handful of publications we store in RMD and using their DOIs with the API, I'm not seeing any that have both.  Seems almost like they remove the Accepted Version once the Published Version is allowed, but I can't be 100% sure.  Maybe if both are there we give preference to Published Version? @Smullz622 @EricDurante

[the remark this answers is not in the document]
Comment (ajkiessl) [reply app/jobs/permissions_check_job.rb:12]: Ah, I spoke too soon.  I found one with both.  In that case the Published Version is deemed the "best" permissions.  Maybe we use the "best" one when both are there.

File: app/models/publication.rb [hunk 2]
```diff
@@ -65,7 +65,11 @@ def self.open_access_statuses
   end
 
   def self.oa_workflow_states
-    ['automatic DOI verification pending']
+    ['automatic DOI verification pending', 'automatic permissions check pending', 'error during permissions check']
+  end
+
+  def self.preferred_versions
+    ['acceptedVersion', 'publishedVersion']
   end
```
Comment (ajkiessl) [inline app/models/publication.rb:73]: We should use the values stored in the translation ymls to set the preferred versions values.  Like how we set valid versions in the oab_permissions_client: https://github.com/psu-libraries/researcher-metadata/blob/fd12910011be99c5931ad208eaa8f7724193ef87/app/services/oab_permissions_client.rb#L7-L8 .

**Review by EricDurante — CHANGES_REQUESTED** [review]
I think that we might need to change a couple of things. Let me know if I've misunderstood anything in my comments.

File: app/services/oab_preferred_permissions_service.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    @permissions = all_permissions
+  end
+
+  def preferred_permission
+    accepted = false
+    published = false
+
+    all_permissions.each do |perm|
+      accepted = true if perm['version'] == I18n.t('file_versions.accepted_version') && perm['can_archive'] == true && perm['locations'].map(&:downcase).include?('institutional repository')
+      published = true if perm['version'] == I18n.t('file_versions.published_version') && perm['can_archive'] == true && perm['locations'].map(&:downcase).include?('institutional repository')
```
[the remark this answers is not in the document]
Comment (Smullz622) [reply app/services/oab_preferred_permissions_service.rb:16]: From what I can tell, it seems that some of the metadata is different, but the metadata we are looking for (license etc) is the same for the same version. My search was far from exhaustive though, so I am not 100% positive if this is always the case.

**Review by EricDurante — APPROVED** [review]

File: app/models/activity_insight_oa_file.rb [hunk 4]
```diff
@@ -3,4 +3,7 @@
 class ActivityInsightOAFile < ApplicationRecord
   belongs_to :publication,
              inverse_of: :activity_insight_oa_files
+
+  scope :pub_without_permissions, -> { left_outer_joins(:publication).where(publication: { licence: nil }).where(publication: { doi_verified: true }) }
+  scope :needs_permissions_check, -> { pub_without_permissions.where(version_checked: nil).where(%{version = 'acceptedVersion' OR version = 'publishedVersion'}) }
```
Comment (ajkiessl) [inline app/models/activity_insight_oa_file.rb:8]: Sorry, I didn't catch this the last time.  Can we do this like this:
```suggestion
  scope :needs_permissions_check, -> { pub_without_permissions.where(version_checked: nil).where(%{version = ? OR version = ?}, I18n.t('file_versions.accepted_version'), I18n.t('file_versions.published_version'))}
```

**Review by ajkiessl — CHANGES_REQUESTED** [review]
This all looks good.  I just have one small change request.

**Review by ajkiessl — APPROVED** [review]

**Comment by ajkiessl** [comment]
I guess we might want to wait until CI starts working again.  I think the new enterprise SSO stuff is preventing it from running? @whereismyjetpack

**Comment by ajkiessl** [comment]
The tests passed locally so I'm going to merge this.  I want to get this merged so I can incorporate it into some of the stuff I'm working on.

**Pull request merged**